File: #1631.path-with-minimum-effort.cpp
class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {

        int row = heights.size();
        int col = heights[0].size();
        const auto dirs = vector<pair<int,int>>{{0,1}, {0, -1}, {-1, 0}, {1, 0}};
        auto visited = vector(row, vector(col, false));
        priority_queue<tuple<int,int,int>> q;
        q.push(tuple{0, 0, 0});
        while (!q.empty()) {
            const auto [eff, i, j] = q.top();
            q.pop();
            if (visited[i][j]) {
                continue;
            }
    
            visited[i][j] = true;
            if (i == row - 1 && j == col - 1) {
                return -eff;
            }

            for (auto [di, dj] : dirs) {
                int x = i + di;
                int y = j + dj;
                if (x < 0 || x >= row || y < 0 || y >= col || visited[x][y]) {
                    continue;
                }
    
                auto min = std::min(eff, - std::abs(heights[x][y] - heights[i][j]));
                q.push(tuple{min, x, y});
            }
        }
        return -1;    
    }
};

File: #1631.path-with-minimum-effort.js
/**
 * @param {number[][]} heights
 * @return {number}
 */
var minimumEffortPath = function(heights) {
    const queue = new MaxPriorityQueue(
        {
            compare: (left, right) => left.effort - right.effort
        }
    );
    queue.enqueue(createRow(0, 0, 0));
    const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];

    const isSafe = (i, j) => {
        if (i < 0 || j < 0 || i >= heights.length || j >= heights[0].length) {
            return false;
        }

        return heights[i][j] < Infinity;
    }

    while (!queue.isEmpty()) {
        const {row, col, effort} = queue.dequeue();
        if (row === heights.length - 1 && col === heights[0].length - 1) {
            return effort;
        }

        const prevValue = heights[row][col];
        heights[row][col] = Infinity;
        for (let direction of directions) {
            const newRow = row + direction[0];
            const newCol = col + direction[1];
            if (isSafe(newRow, newCol)) {
                const dif = Math.abs(heights[newRow][newCol] - prevValue);
                const newEffort = Math.max(dif, effort);
                queue.enqueue(createRow(newRow, newCol, newEffort));
            }
        }
    }

    return 0;
};
const createRow = (row, col, effort) => ({ row, col, effort });
